docs/proposals/20200220-cluster-resource-set.md

```diff
@@ -0,0 +1,202 @@
+
+#### Story 2
+
+As someone creating multiple clusters, I want all my clusters to have a StorageClass installed automatically, so I don't have to manually repeat the installation for each new cluster.
+
+#### Story 3
+
+As someone creating multiple clusters, I want to be able to provide different values for some fields in the resources for different clusters. For example, CNIs podCIDRs  may be required to be distinct for each cluster, hence some templating mechanism for variable substitution in the resources is needed.
+
+### Implementation Details/Notes/Constraints
+
+#### Data model changes
+
+None. We are planning to implement this feature without modifying any of the existing structure to minimize the footprint of ClusterResourceSet Controller. This enhancement will follow Kubernetes’s feature-gate structure and will be under the experimental package with its APIs, and enabled/disabled with a feature gate. 
+
+#### ClusterResourceSet Object Definition
+
+This is the CRD that is used to have a set of components that will be applied to clusters that match the label selector in it.
+
+The resources field is a list of secrets/configmaps in the same namespace. As a cluster label selector, any key-value pair works as long as the same key-value label is assigned to clusters that the addon will be applied to. The reason not to use a predefined label here is to allow matching with multiple ClusterResourceSet objects.
+
+*Sample ClusterResourceSet YAML*
+
+```yaml
+---
+apiVersion: cluster.x-k8s.io/v1alpha3
+kind: ClusterResourceSet
+metadata:
+ name: postcreate-conf
+ namespace: default
+spec:
+ mode: "ApplyOnce" / "Sync"
+ clusterSelector:
+   matchLabels:
+     postcreatelabelcni: calico
+ resources:
+   - name: calico-addon
+     kind: Secret
+   - name: network-policy-addon
+     kind: ConfigMap
+```
+
+Initially, the only mode supported will be "ApplyOnce" and it will be the default mode if no mode is provided. "Sync" mode will include reapplying the resources on resource hash change or periodically following an interval.
+Each item in the resources specifies a kind (must be either ConfigMap or Secret) and a name. Each referenced ConfigMap/Secret  contains yaml/json content as value. The key to that content is “value”.
+
+*Sample Secret Format*
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name:calico-addon
+type: Opaque
+stringData:
+  value: |-
+    kind: ConfigMap
+    apiVersion: v1
+    metadata:
+     name: calico-conf
+```
+
+The resources in ClusterResourceSet will be applied to matching clusters. 
+A configmap will be created in the management cluster to keep track of which resources are applied by ClusterResourceSet resources. There will be one configmap per workload cluster.
+
+Example:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: capi-crs-my-awesome-cluster
+  namespace: cluster-api-system
```

Comment:
Will this match the namespace of the Cluster?

Reply:
I was thinking to have it in the same namespaces as ClusterResourceSet so that all ClusterResourceSet related components (resources and ConfigMaps) can be under same ns. WDYT?

Reply:
I think the format of this ConfigMap is one per cluster, listing the status of every ClusterResourceSet that applies to the cluster. Right?

Reply:
Correct. I am fine with putting them in Cluster's namespace.